File: depot_planner/hybrid/reeds_shepp.py
"""Reeds-Shepp curves: shortest paths for a car that can drive forwards and back.

Used as the analytic expansion of hybrid A*: once the search gets near the goal,
a Reeds-Shepp curve connects the last state to the goal pose exactly, which is
what makes the 0.3 m / 5 degree goal tolerance reachable at all.

The word set covers CSC (``LSL``, ``LSR``), CCC (``LRL``) and SCS (``SLS``)
families, each expanded by the standard timeflip and reflect symmetries. That is
not the full 48-word Reeds-Shepp set, so the curve returned is a valid shortest
*candidate*, not provably the global optimum; every candidate is collision
checked and its endpoint verified before use.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Callable, Sequence

from depot_planner.hybrid.car import CarModel, Pose, wrap_angle

LEFT, STRAIGHT, RIGHT = 1, 0, -1


@dataclass(frozen=True)
class RSSegment:
    """One Reeds-Shepp segment: a steering direction and a signed length.

    ``length`` is in units of the turning radius; its sign is the gear
    (positive forward, negative reverse).
    """

    steering: int
    length: float


@dataclass(frozen=True)
class RSPath:
    """A sequence of Reeds-Shepp segments."""

    segments: tuple[RSSegment, ...]

    @property
    def length(self) -> float:
        """Total path length in units of the turning radius."""
        return sum(abs(segment.length) for segment in self.segments)

    @property
    def n_direction_changes(self) -> int:
        gears = [1 if s.length >= 0.0 else -1 for s in self.segments if abs(s.length) > 1e-9]
        return sum(1 for a, b in zip(gears, gears[1:]) if a != b)


def _polar(x: float, y: float) -> tuple[float, float]:
    return math.hypot(x, y), math.atan2(y, x)


def _mod2pi(theta: float) -> float:
    """Wrap to ``(-pi, pi]`` (the convention the word formulas below assume)."""
    value = math.fmod(theta, 2.0 * math.pi)
    if value < -math.pi:
        value += 2.0 * math.pi
    elif value > math.pi:
        value -= 2.0 * math.pi
    return value


# ------------------------------------------------------------- the base words


def _lsl(x: float, y: float, phi: float) -> tuple[float, float, float] | None:
    u, t = _polar(x - math.sin(phi), y - 1.0 + math.cos(phi))
    if 0.0 <= t <= math.pi:
        v = _mod2pi(phi - t)
        if 0.0 <= v <= math.pi:
            return t, u, v
    return None


def _lsr(x: float, y: float, phi: float) -> tuple[float, float, float] | None:
    u1, t1 = _polar(x + math.sin(phi), y - 1.0 - math.cos(phi))
    squared = u1 * u1
    if squared >= 4.0:
        u = math.sqrt(squared - 4.0)
        theta = math.atan2(2.0, u)
        t = _mod2pi(t1 + theta)
        v = _mod2pi(t - phi)
        if t >= 0.0 and v >= 0.0:
            return t, u, v
    return None


def _lrl(x: float, y: float, phi: float) -> tuple[float, float, float] | None:
    u1, t1 = _polar(x - math.sin(phi), y - 1.0 + math.cos(phi))
    if u1 <= 4.0:
        u = -2.0 * math.asin(0.25 * u1)
        t = _mod2pi(t1 + 0.5 * u + math.pi)
        v = _mod2pi(phi - t + u)
        if t >= 0.0 and u <= 0.0:
            return t, u, v
    return None


def _sls(x: float, y: float, phi: float) -> tuple[float, float, float] | None:
    phi = _mod2pi(phi)
    if abs(y) < 1e-9 or not (1e-9 < abs(phi) < math.pi * 0.99):
        return None
    if phi < 0.0:
        return None
    xd = -y / math.tan(phi) + x
    t = xd - math.tan(phi / 2.0)
    u = phi
    v = math.hypot(x - xd, y) - math.tan(phi / 2.0)
    if y < 0.0:
        v = -math.hypot(x - xd, y) - math.tan(phi / 2.0)
    return t, u, v


def _mirror(word: Sequence[int]) -> tuple[int, ...]:
    return tuple(-s for s in word)


#: (base function, steering word) pairs; each is expanded by four symmetries.
_WORDS: tuple[tuple[Callable[..., tuple[float, float, float] | None], tuple[int, int, int]], ...] = (
    (_lsl, (LEFT, STRAIGHT, LEFT)),
    (_lsr, (LEFT, STRAIGHT, RIGHT)),
    (_lrl, (LEFT, RIGHT, LEFT)),
    (_sls, (STRAIGHT, LEFT, STRAIGHT)),
)


def _candidates(x: float, y: float, phi: float) -> list[RSPath]:
    """Every candidate curve for a goal at ``(x, y, phi)`` with unit turning radius."""
    paths: list[RSPath] = []

    def add(lengths: tuple[float, float, float], word: Sequence[int]) -> None:
        if any(not math.isfinite(value) for value in lengths):
            return
        segments = tuple(
            RSSegment(steering, length)
            for steering, length in zip(word, lengths)
            if abs(length) > 1e-9
        )
        if segments:
            paths.append(RSPath(segments))

    for base, word in _WORDS:
        mirrored = _mirror(word)
        result = base(x, y, phi)
        if result is not None:
            add(result, word)
        result = base(-x, y, -phi)                      # timeflip
        if result is not None:
            add(tuple(-value for value in result), word)
        result = base(x, -y, -phi)                      # reflect
        if result is not None:
            add(result, mirrored)
        result = base(-x, -y, phi)                      # timeflip + reflect
        if result is not None:
            add(tuple(-value for value in result), mirrored)
    return paths


def reeds_shepp_paths(start: Pose, goal: Pose, max_curvature: float) -> list[RSPath]:
    """All candidate Reeds-Shepp curves from ``start`` to ``goal``."""
    dx = goal[0] - start[0]
    dy = goal[1] - start[1]
    cos_t, sin_t = math.cos(start[2]), math.sin(start[2])
    x = (cos_t * dx + sin_t * dy) * max_curvature
    y = (-sin_t * dx + cos_t * dy) * max_curvature
    phi = wrap_angle(goal[2] - start[2])
    return _candidates(x, y, phi)


def shortest_reeds_shepp(start: Pose, goal: Pose, max_curvature: float) -> RSPath | None:
    """The shortest candidate curve, or ``None`` if no word applies."""
    paths = reeds_shepp_paths(start, goal, max_curvature)
    return min(paths, key=lambda p: p.length) if paths else None


def interpolate(
    start: Pose, path: RSPath, car: CarModel, step: float = 0.2
) -> tuple[list[Pose], list[int]]:
    """Sample a curve into poses and per-step gears (``+1`` forward, ``-1`` reverse).

    The first pose returned is ``start``; ``directions[i]`` is the gear used to
    get from ``poses[i]`` to ``poses[i + 1]``.
    """
    radius = car.min_turning_radius
    poses: list[Pose] = [start]
    directions: list[int] = []
    pose = start
    for segment in path.segments:
        distance_m = abs(segment.length) * radius
        gear = 1 if segment.length >= 0.0 else -1
        steer = segment.steering * car.max_steer
        substeps = max(1, int(math.ceil(distance_m / step)))
        piece = gear * distance_m / substeps
        for _ in range(substeps):
            pose = car.step(pose, steer, piece)
            poses.append(pose)
            directions.append(gear)
    return poses, directions


def path_length_m(path: RSPath, car: CarModel) -> float:
    return path.length * car.min_turning_radius

File: cpp/src/reeds_shepp.cpp
#include "depot/reeds_shepp.hpp"

#include <cmath>

#include "depot/numeric.hpp"

namespace depot {
namespace {

/// One word's solution: three signed lengths, or "no solution".
struct Lengths {
  bool ok = false;
  double t = 0.0;
  double u = 0.0;
  double v = 0.0;
};

void Polar(double x, double y, double* radius, double* angle) {
  *radius = PythonHypot(x, y);
  *angle = std::atan2(y, x);
}

Lengths Lsl(double x, double y, double phi) {
  double u, t;
  Polar(x - std::sin(phi), y - 1.0 + std::cos(phi), &u, &t);
  if (t >= 0.0 && t <= kPi) {
    const double v = Mod2Pi(phi - t);
    if (v >= 0.0 && v <= kPi) return Lengths{true, t, u, v};
  }
  return Lengths{};
}

Lengths Lsr(double x, double y, double phi) {
  double u1, t1;
  Polar(x + std::sin(phi), y - 1.0 - std::cos(phi), &u1, &t1);
  const double squared = u1 * u1;
  if (squared >= 4.0) {
    const double u = std::sqrt(squared - 4.0);
    const double theta = std::atan2(2.0, u);
    const double t = Mod2Pi(t1 + theta);
    const double v = Mod2Pi(t - phi);
    if (t >= 0.0 && v >= 0.0) return Lengths{true, t, u, v};
  }
  return Lengths{};
}

Lengths Lrl(double x, double y, double phi) {
  double u1, t1;
  Polar(x - std::sin(phi), y - 1.0 + std::cos(phi), &u1, &t1);
  if (u1 <= 4.0) {
    const double u = -2.0 * std::asin(0.25 * u1);
    const double t = Mod2Pi(t1 + 0.5 * u + kPi);
    const double v = Mod2Pi(phi - t + u);
    if (t >= 0.0 && u <= 0.0) return Lengths{true, t, u, v};
  }
  return Lengths{};
}

Lengths Sls(double x, double y, double phi) {
  phi = Mod2Pi(phi);
  if (std::fabs(y) < 1e-9 || !(std::fabs(phi) > 1e-9 && std::fabs(phi) < kPi * 0.99)) {
    return Lengths{};
  }
  if (phi < 0.0) return Lengths{};
  const double xd = -y / std::tan(phi) + x;
  const double t = xd - std::tan(phi / 2.0);
  const double u = phi;
  double v = PythonHypot(x - xd, y) - std::tan(phi / 2.0);
  if (y < 0.0) v = -PythonHypot(x - xd, y) - std::tan(phi / 2.0);
  return Lengths{true, t, u, v};
}

using WordFunction = Lengths (*)(double, double, double);

struct Word {
  WordFunction base;
  int steering[3];
};

/// (base function, steering word) pairs; each is expanded by four symmetries,
/// in exactly the order _WORDS and _candidates list them.
const Word kWords[4] = {
    {&Lsl, {kLeft, kStraight, kLeft}},
    {&Lsr, {kLeft, kStraight, kRight}},
    {&Lrl, {kLeft, kRight, kLeft}},
    {&Sls, {kStraight, kLeft, kStraight}},
};

void Add(std::vector<RSPath>* paths, const Lengths& lengths, const int steering[3]) {
  const double values[3] = {lengths.t, lengths.u, lengths.v};
  for (double value : values) {
    if (!std::isfinite(value)) return;
  }
  RSPath path;
  for (int i = 0; i < 3; ++i) {
    if (std::fabs(values[i]) > 1e-9) path.segments.push_back(RSSegment{steering[i], values[i]});
  }
  if (!path.segments.empty()) paths->push_back(std::move(path));
}

Lengths Negated(const Lengths& lengths) {
  return Lengths{lengths.ok, -lengths.t, -lengths.u, -lengths.v};
}

}  // namespace

int RSPath::DirectionChanges() const {
  std::vector<int> gears;
  for (const RSSegment& segment : segments) {
    if (std::fabs(segment.length) > 1e-9) gears.push_back(segment.length >= 0.0 ? 1 : -1);
  }
  int changes = 0;
  for (std::size_t i = 0; i + 1 < gears.size(); ++i) {
    if (gears[i] != gears[i + 1]) ++changes;
  }
  return changes;
}

std::vector<RSPath> ReedsSheppPaths(const Pose& start, const Pose& goal, double max_curvature) {
  const double dx = goal.x - start.x;
  const double dy = goal.y - start.y;
  const double cos_t = std::cos(start.theta);
  const double sin_t = std::sin(start.theta);
  const double x = (cos_t * dx + sin_t * dy) * max_curvature;
  const double y = (-sin_t * dx + cos_t * dy) * max_curvature;
  const double phi = WrapAngle(goal.theta - start.theta);

  std::vector<RSPath> paths;
  for (const Word& word : kWords) {
    const int mirrored[3] = {-word.steering[0], -word.steering[1], -word.steering[2]};

    Lengths result = word.base(x, y, phi);
    if (result.ok) Add(&paths, result, word.steering);

    result = word.base(-x, y, -phi);  // timeflip
    if (result.ok) Add(&paths, Negated(result), word.steering);

    result = word.base(x, -y, -phi);  // reflect
    if (result.ok) Add(&paths, result, mirrored);

    result = word.base(-x, -y, phi);  // timeflip + reflect
    if (result.ok) Add(&paths, Negated(result), mirrored);
  }
  return paths;
}

RSPath ShortestReedsShepp(const Pose& start, const Pose& goal, double max_curvature, bool* found) {
  const std::vector<RSPath> paths = ReedsSheppPaths(start, goal, max_curvature);
  if (paths.empty()) {
    *found = false;
    return RSPath{};
  }
  // Python's min() keeps the first of several equal minima, so the comparison
  // has to be strictly less-than and the scan has to run in order.
  std::size_t best = 0;
  double best_length = paths[0].Length();
  for (std::size_t i = 1; i < paths.size(); ++i) {
    const double length = paths[i].Length();
    if (length < best_length) {
      best = i;
      best_length = length;
    }
  }
  *found = true;
  return paths[best];
}

void InterpolateReedsShepp(const Pose& start, const RSPath& path, const CarModel& car, double step,
                           std::vector<Pose>* poses, std::vector<int>* directions) {
  const double radius = car.min_turning_radius();
  poses->clear();
  directions->clear();
  poses->push_back(start);
  Pose pose = start;
  for (const RSSegment& segment : path.segments) {
    const double distance_m = std::fabs(segment.length) * radius;
    const int gear = segment.length >= 0.0 ? 1 : -1;
    const double steer = segment.steering * car.max_steer;
    const int substeps = std::max(1, static_cast<int>(std::ceil(distance_m / step)));
    const double piece = gear * distance_m / substeps;
    for (int i = 0; i < substeps; ++i) {
      pose = car.Step(pose, steer, piece);
      poses->push_back(pose);
      directions->push_back(gear);
    }
  }
}

double ReedsSheppLengthMetres(const RSPath& path, const CarModel& car) {
  return path.Length() * car.min_turning_radius();
}

}  // namespace depot
